src/app.js

```diff
@@ -25,6 +25,36 @@ if ( !components ) {
 
 components.widget.toggle( true );
 components.widget.setState( 'pending' );
+components.api.onError( function ( errorCode ) {
+	components.widget.setState( 'err' );
+	components.widget.setErrorMessage( errorCode );
+} );
+components.api.onResult( function ( result ) {
+	// Insert modified HTML.
+	$( '.mw-parser-output' ).replaceWith( $( result.replacementHtml ) );
+	// Highlight when hover a user's contributions.
+	// @TODO This is just testing code and should be moved and replaced by the proper behaviour.
+	$( '.mw-parser-output .editor-token' ).hover(
+		function () {
+			let editorIdMatches, userId;
+			// Get the editor ID from the span's class name.
+			editorIdMatches = $( this ).attr( 'class' ).match( /token-editor-(\d+)/ );
+			if ( !editorIdMatches || editorIdMatches[ 1 ] === undefined ) {
+				// Ignore any other '.editor-token' elements.
+				return;
+			}
+			userId = editorIdMatches[ 1 ];
+			$( '.token-editor-' + userId ).addClass( 'active' );
+			// Revisions are: result.revisionsByUser[ userId ]
+			// and user info is: result.users[ userId ]
+		},
+		function () {
+			$( '.mw-parser-output .editor-token' ).removeClass( 'active' );
+		}
+	);
+	components.widget.setState( 'ready' );
+} );
+components.api.getData( window.location.href );
```

Comment:
To continue my point from the comment above, if we assume `getData` always returns a promise that is resolved or rejected based on errors, the flow of this code would also look nicer.

```
components.api.getData( ... )
   .then( ... ) // success
   .fail( ... ) // failure
```
Or it would also let us chain responses; for example, do something like 

```
showSpinner()
api.getData()
    .then( ... )
    .fail( ... )
    .always( () => hideSpinner );
```

Reply:
Done.